### PR TITLE
Stream support

### PR DESCRIPTION
**Pull request by telphan**

### CODE REVIEW

File: lib/helpers/index.ts [hunk 1]
```diff
@@ -1,8 +1,12 @@
-import * as Bluebird from 'bluebird';
-import * as path from 'path';
-import * as express from 'express';
 import * as sdk from 'etcher-sdk';
 
+export function promiseStream(stream: NodeJS.WritableStream): Promise<void> {
```
Comment (CameronDiver) [inline lib/helpers/index.ts:3]:
```suggestion
export function promiseStream(stream: Stream.Stream): Promise<void> {
```
With `import * as Stream from 'stream';` above.

File: lib/index.ts [hunk 2]
```diff
@@ -29,19 +29,26 @@ async function setup(devicePath: string, port: number = 80): Promise<Server> {
...
 		res.send('OK');
-		await worker.off();
 	});
 	app.post(
 		'/dut/flash',
 		async (req: express.Request, res: express.Response) => {
-			res.status(202).send('In Progress');
-			await worker.flash(req.body.url);
+			function keepAlive() {
+				res.write('Still flashing');
```
Comment (CameronDiver) [inline lib/index.ts:44]: you're probably better off doing
```
const timer = setInterval(() => {
  res.write('Still flashing');
}, keepaliveTimeout)
```

and then use the `timer` handle to clear the interval when it has finished processing.

File: lib/workers/testbot.ts [hunk 3]
```diff
@@ -161,44 +162,73 @@ export class TestBot {
...
+			});
+		}, arguments);
 	}
 
 	/**
 	 * Flash SD card with operating system
 	 */
-	public async flash(url: string): Promise<void> {
-		await this.off();
+	public async flashDUT(stream: NodeJS.ReadableStream): Promise<void> {
```
Comment (CameronDiver) [inline lib/workers/testbot.ts:195]:
```suggestion
	public async flashDUT(stream: Stream.Readable): Promise<void> {
```

File: lib/workers/testbot.ts [hunk 4]
```diff
@@ -161,44 +162,73 @@ export class TestBot {
...
+					);
+					const progressBar: { [key: string]: any } = {
+						flashing: new visuals.Progress('Flashing'),
+						verifying: new visuals.Progress('Validating'),
+					};
+
+					await sdk.multiWrite.pipeSourceToDestinations(
+						source,
+						[drive],
+						(destination, error) => {
```
Comment (CameronDiver) [inline lib/workers/testbot.ts:219]:
```suggestion
						(_destination, error) => {
```

File: test/helpers.spec.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import 'mocha';
+import * as chai from 'chai';
+import * as chaiAsPromised from 'chai-as-promised';
+import * as helpers from '../lib/helpers';
+import { join } from 'path';
+import { fs } from 'mz';
+import { Writable } from 'readable-stream';
```
Comment (CameronDiver) [inline test/helpers.spec.ts:7]: This isn't in your package.json, but also looks like it isn't needed?

File: test/helpers.spec.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+import { callback } from 'serialport';
+
+chai.use(chaiAsPromised);
+const { expect } = chai;
+
+class DevNull extends Writable {
```
Comment (CameronDiver) [inline test/helpers.spec.ts:13]: his can just be `const DevNull = fs.createWritableStream('/dev/null')`

File: test/helpers.spec.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+		super();
+	}
+
+	_write(chunk: never, encoding: never, cb: callback) {
+		setImmediate(cb);
+	}
+}
+
+describe('promiseStream', () => {
+	let stream: NodeJS.WritableStream;
```
Comment (CameronDiver) [inline test/helpers.spec.ts:24]:
```suggestion
	let stream: Stream.Writable;
```

**Comment by telphan** [comment]
#12 supersedes this PR.

**Pull request closed without merging**